store/multistore/store.go

```diff
@@ -0,0 +1,301 @@
+	"strings"
+	"time"
+
+	"cosmossdk.io/log"
+	v1types "cosmossdk.io/store/types"
+	"cosmossdk.io/store/v2"
+	"cosmossdk.io/store/v2/commitment"
+	"github.com/cockroachdb/errors"
+	ics23 "github.com/cosmos/ics23/go"
+)
+
+type (
+	// MultiStore defines an abstraction layer containing a State Storage (SS) engine
+	// and one or more State Commitment (SC) engines.
+	//
+	// TODO:
+	// - Move relevant types to the 'core' package.
+	// - Remove reliance on store v1 types.
+	MultiStore interface {
+		GetSCStore(storeKey string) *commitment.Database
+		MountSCStore(storeKey string, sc *commitment.Database) error
+		GetProof(storeKey string, version uint64, key []byte) (*ics23.CommitmentProof, error)
+		LoadVersion(version uint64) error
+		LoadLatestVersion() error
+		GetLatestVersion() (uint64, error)
+		WorkingHash() []byte
+		Commit() ([]byte, error)
+		SetCommitHeader(h CommitHeader)
+
+		// TODO:
+		// - Tracing
+		// - Branching
+		// - Queries
+
+		io.Closer
+	}
+
+	CommitHeader interface {
+		GetTime() time.Time
+		GetHeight() uint64
+	}
+)
+
+var _ MultiStore = &Store{}
+
+type Store struct {
+	logger         log.Logger
+	commitHeader   CommitHeader
+	initialVersion uint64
+
+	// ss reflects the state storage backend
+	ss store.VersionedDatabase
+
+	// scStores reflect a mapping of store key to state commitment backend (i.e. a backend per module)
+	scStores map[string]*commitment.Database
+
+	// removalMap reflects module stores marked for removal
+	removalMap map[string]struct{}
+
+	// lastCommitInfo reflects the last version/hash that has been committed
+	lastCommitInfo *v1types.CommitInfo
+}
+
+func New(logger log.Logger, initialVersion uint64, ss store.VersionedDatabase) (MultiStore, error) {
+	return &Store{
+		logger:         logger.With("module", "multi_store"),
+		initialVersion: initialVersion,
+		ss:             ss,
+		scStores:       make(map[string]*commitment.Database),
+		removalMap:     make(map[string]struct{}),
+	}, nil
+}
+
+// Close closes the store and resets all internal fields. Note, Close() is NOT
+// idempotent and should only be called once.
+func (s *Store) Close() (err error) {
+	err = errors.Join(err, s.ss.Close())
+	for _, sc := range s.scStores {
+		err = errors.Join(err, sc.Close())
+	}
+
+	s.ss = nil
+	s.scStores = nil
+	s.lastCommitInfo = nil
+	s.commitHeader = nil
+	s.removalMap = nil
+
+	return err
+}
+
+func (s *Store) MountSCStore(storeKey string, sc *commitment.Database) error {
+	s.logger.Debug("mounting store", "store_key", storeKey)
+	if _, ok := s.scStores[storeKey]; ok {
+		return fmt.Errorf("SC store with key %s already mounted", storeKey)
+	}
+
+	s.scStores[storeKey] = sc
+	return nil
+}
+
+// LastCommitID returns a CommitID based off of the latest internal CommitInfo.
+// If an internal CommitInfo is not set, a new one will be returned with only the
+// latest version set, which is based off of the SS view.
+func (s *Store) LastCommitID() (v1types.CommitID, error) {
+	if s.lastCommitInfo == nil {
```

Comment:
We can invert this condition, have an early return then reduce nesting and make it more implicit to read, like this
```go
if sci := s.lastCommitInfo; sci != nil {
    return sci.CommitID(), nil
}

lv, err := s.ss.GetLatestVersion()
if err != nil {
    return v1types.CommitID{}, err
}
...
```